Modularize embeddings

mberr opened the pull request:
This PR introduces a modularization of entity/relation representation providers.
(as already discussed with @mali-git @cthoyt )

This offers multiple benefits:
* easier use of custom representations (e.g. obtained as transformation of initial embeddings, from other modalities, etc.)
* easier configuration of embedding initializations & constraints, and its inclusion in HPO searches.
* in combination with #107 , this allows to easier match the "model" definition with the combination mentioned in the [paper](https://arxiv.org/abs/2006.13365).

Tasks:
* [x] move initialization/reset_parameters into the `Embedding` module
* [x] move constraints inside the `Embedding` module
* [x] replace `nn.Embedding.weight` use by `Embedding.forward()`

mberr committed:
Add base module for representations

mberr committed:
Add reset_parameters

mberr committed:
Add wrapper for nn.Embedding

mberr committed:
Implement forward

mberr committed:
Add properties

mberr committed:
Use Embedding for EntityRelationEmbedding

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -183,6 +183,72 @@ def _process_remove_known(df: pd.DataFrame, remove_known: bool, testing: Optiona
+        )
+
+    @property
+    def total_size(self) -> int:  # noqa: D102
```
Why not name this num_embeddings to reflect the nn.Embedding?

mberr replied:
Maybe `max_id` would be better.

I think "representation" is a much broader term than "embedding", where the former one can be obtained from anything, e.g. via a CNN from an image, whereas the latter is usually a trained vector.

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1078,12 +1144,11 @@ def __init__(
+        self.entity_embeddings = Embedding(
+            num=triples_factory.num_entities,
+            dim=self.embedding_dim,
+        ).to(device=self.device)
```
Why not make it possible to give the device directly in the new class's init to better match the old interface?

mberr replied:
I actually experienced initialization to be significantly faster when always done on cpu and then moved to GPU.

cthoyt replied:
Really interesting insight!

mberr replied:
for a discussion about some of the reasons you might look into https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-37-efficient-random-number-generation-and-application

cthoyt reviewed (commented):
Really nice idea max. See comments above

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -19,7 +19,7 @@
 from ..tqdmw import tqdm
 from ..triples import TriplesFactory
 from ..typing import MappedTriples
-from ..utils import NoRandomSeedNecessary, get_embedding, resolve_device, set_random_seed
```
Is the get_embedding function not used anymore? Maybe it should delegate to the new code and emit a deprecation warning

mberr replied:
It is used in a few places.

mberr committed:
Replace old factory method

cthoyt committed:
Reorganize

cthoyt committed:
Revert for now

cthoyt committed:
Merge branch 'master' into modularize-embeddings

cthoyt commented on src/pykeen/models/init.py:
```diff
@@ -31,7 +31,8 @@ def embedding_xavier_uniform_(embedding: nn.Embedding, gain: float = 1.) -> nn.E
 
-    >>> e = nn.Embedding(num_embeddings=3, embedding_dim=5)
+    >>> import pykeen.utils
+    >>> e = pykeen.utils.Embedding(num_embeddings=3, embedding_dim=5)
```
should we have a `pykeen.nn` module so it seems more like a drop-in replacement?

mberr replied:
Sure, it might not be well-placed in `utils` anyway

cthoyt replied:
Done in aeda9b89

cthoyt commented on src/pykeen/utils.py:
```diff
@@ -343,44 +345,131 @@ def save_to_s3(self, directory: str, bucket: str, s3=None) -> None:
-        The embedding.
-    """
-    # Allocate weight on device
-    weight = torch.empty(num_embeddings, embedding_dim, device=device)
```
why don't we have a `weight` in the new implementation?

mberr replied:
We needed it before, since nn.Embedding initializes on CPU, if we do not explicitly provide a weight already living on GPU. Since we now want to initialize on cpu anyway, we can again use the default implementation from torch.

cthoyt commented on src/pykeen/utils.py:
```diff
@@ -343,44 +345,131 @@ def save_to_s3(self, directory: str, bucket: str, s3=None) -> None:
+        else:
+            self.initialization = initialization
+        self.normalization = normalization
+        self._embeddings = nn.Embedding(
```
why no weights here? also see https://github.com/pykeen/pykeen/pull/132/files#r518005303

cthoyt committed:
Make initialization functions work

cthoyt committed:
Update utils.py

cthoyt committed:
Reorganize again

cthoyt committed:
Fix docs

cthoyt committed:
Update base.py

cthoyt committed:
Fix initialization

mberr committed:
Update RepresentationModel.forward API

mberr committed:
Add deprecation warning

mberr committed:
Add post_parameter_update

mberr committed:
Add constrainer

mberr committed:
Forward kwargs from EntityRelationEmbeddingModel

mberr committed:
Use Embedding.init_with_device for EntityEmbeddingModel

mberr committed:
Use new-style Embeddings for DistMult

mberr committed:
Fix delegation of post_parameter_update

cthoyt committed:
Update code style

mberr committed:
Add unittest to check whether the model can handle custom representations

mberr committed:
Merge remote-tracking branch 'origin/modularize-embeddings' into modularize-embeddings

cthoyt committed:
Pass flake8

cthoyt committed:
New implemntations of init functions

cthoyt committed:
Fix attribute names

cthoyt committed:
Add first round of upgrades for _reset_parameters_

cthoyt committed:
Merge branch 'master' into modularize-embeddings

cthoyt committed:
Switch more over

cthoyt commented on src/pykeen/models/unimodal/distmult.py:
```diff
@@ -146,9 +138,9 @@ def score_hrt(self, hrt_batch: torch.LongTensor) -> torch.FloatTensor:  # noqa:
-        h = self.entity_embeddings(hr_batch[:, 0]).view(-1, 1, self.embedding_dim)
-        r = self.relation_embeddings(hr_batch[:, 1]).view(-1, 1, self.embedding_dim)
-        t = self.entity_embeddings.weight.view(1, -1, self.embedding_dim)
+        h = self.entity_embeddings(indices=hr_batch[:, 0]).view(-1, 1, self.embedding_dim)
```
@mberr  is the keyword argument necessary?

mberr replied:
No. I generally prefer to use keyword whenever possible since it is more robust against changing the order of parameters.

cthoyt replied:
I agree in almost all cases, except when it's single argument functions

cthoyt committed:
Update utils.py

cthoyt commented on src/pykeen/utils.py:
```diff
@@ -203,7 +206,7 @@ def _flatten_dictionary(
 
 
 def get_embedding_in_canonical_shape(
```
This function is never called in PyKEEN with `ind=None`. Is there a reason it has this functionality implemented?

cthoyt committed:
Add TODOs

[skip ci]

cthoyt committed:
Make more general

cthoyt committed:
Enable more kwargs

cthoyt committed:
Fix up KG2E

mberr committed:
Fix unittest

mberr committed:
Simplify get_embedding_in_canonical_shape

mberr committed:
Directly use Embedding.init_with_device

mberr committed:
Replace deprecated usage for ConvE

mberr committed:
Use new-style embeddings for TransD

mberr committed:
Add in-place norm clamping

mberr committed:
Fix unittest for TransD; remove manual score comparison

cthoyt committed:
Fix docs

mberr committed:
Merge remote-tracking branch 'origin/modularize-embeddings' into modularize-embeddings

# Conflicts:
#	src/pykeen/utils.py

mberr committed:
Forward kwargs

mberr committed:
Merge remote-tracking branch 'origin/modularize-embeddings' into modularize-embeddings

mberr committed:
More generic types

mberr committed:
More generic types, now as TypeVar

mberr committed:
rename parameters

mberr committed:
fix renamed parameters

mberr committed:
Add comments

mberr committed:
Remove deprecated .weight property

mberr committed:
Fix ConvE's weight usage

mberr committed:
Fix .weight usage in test_save_load_model_state

mberr committed:
reduce code duplication in test

mberr committed:
do not require in-place initializer

mberr committed:
Remove manual unittest for TransR

cthoyt commented on src/pykeen/models/unimodal/trans_d.py:
```diff
@@ -124,38 +125,25 @@ def __init__(
             regularizer=regularizer,
+            entity_initializer=xavier_normal_,
+            relation_initializer=xavier_normal_,
+            entity_constrainer=functools.partial(clamp_norm_, maxnorm=1., p=2, dim=-1),
```
Would you rather do this than use entity_constrainer_kwargs?

mberr committed:
Fix .weight usage in test_models.py

mberr committed:
Remove post_parameter_update tests after embeddings have been exchanged

mberr committed:
Add TODO to complex

mberr committed:
use indices

mberr committed:
Remove no_grad annotation, since it was causing problems with general gradient tracking

mberr committed:
cleanup conve

mberr committed:
cleanup convkb

mberr committed:
cleanup distmult

mberr committed:
Use keyword parameter, fix .weight

mberr committed:
Use keyword parameter, fix .weight

mberr committed:
Add constrainer, fix .weight

mberr committed:
fix .weight

mberr committed:
fix .weight

mberr committed:
fix .weight, use initializer and constrainer

mberr committed:
Directly use Embedding.init_with_device; fix type annotation

mberr committed:
use new-style embeddings

mberr committed:
use *_kwargs instead of functools.partial

mberr committed:
Use new-style embeddings in TransE

mberr committed:
Use new-style embeddings in TransR

mberr committed:
Use new-style embeddings in Tucker

mberr committed:
Use new-style embeddings in UM

mberr committed:
Add chain_ and normalize_ utilities

mberr committed:
Fix chain_

mberr committed:
Fix complex special case for hrt

mberr committed:
fix custom representation unittest

mberr committed:
Add missing calls to reset_parameters

mberr committed:
Fix tests for get_embedding_in_canonical_shape

mberr committed:
Use magicmock to check whether reset_parameters was called

mberr committed:
hotfix RGCN

nice fixes need
https://github.com/pykeen/pykeen/pull/107
https://github.com/pykeen/pykeen/pull/110

mberr committed:
Fix block decomposition

mberr committed:
Fix TransR

mberr committed:
Add trailing comma

mberr committed:
Fix line too long

mberr committed:
Add trailing comma

mberr committed:
Fix line-too-long

mberr committed:
Fix wrong comparison

mberr committed:
Remove trailing comma

mberr committed:
fix line too long

mberr committed:
fix line too long

mberr committed:
fix line too long

cthoyt commented on src/pykeen/models/unimodal/hole.py:
```diff
@@ -71,20 +71,13 @@ def __init__(
+            # Initialisation, cf. https://github.com/mnick/scikit-kge/blob/master/skge/param.py#L18-L27
+            entity_initializer=xavier_uniform_,
+            relation_initializer=xavier_uniform_,
+            entity_constrainer=clamp_norm,
```
should we switch this to in-place?

cthoyt replied:
okay. we just have to keep straight that initializer functions have to be in place, while constrainer and normalizer ones have to be out place

mberr replied:
The init does not need to be in-place, in case we take care of operating in-place, cf. the implementation of torch.nn.init.*, e.g. https://pytorch.org/docs/stable/_modules/torch/nn/init.html#uniform_

cthoyt commented on src/pykeen/models/unimodal/kg2e.py:
```diff
@@ -83,6 +84,10 @@ def __init__(
             preferred_device=preferred_device,
             random_seed=random_seed,
             regularizer=regularizer,
+            entity_constrainer=clamp_norm,
```
should we switch to in-place?

cthoyt commented on src/pykeen/models/unimodal/kg2e.py:
```diff
@@ -98,15 +103,21 @@ def __init__(
             embedding_dim=embedding_dim,
             device=self.device,
+            # Ensure positive definite covariances matrices and appropriate size by clamping
+            constrainer=torch.clamp,
```
this was clamp norm before

cthoyt replied:
ooops no it wasn't

cthoyt commented on src/pykeen/models/unimodal/trans_e.py:
```diff
@@ -73,42 +74,38 @@ def __init__(
+            entity_initializer=xavier_uniform_,
+            relation_initializer=chain_(
+                xavier_uniform_,
+                functional.normalize,
```
should this be in-place be `normalize_`?

cthoyt commented on src/pykeen/models/unimodal/trans_h.py:
```diff
@@ -90,47 +90,39 @@ def __init__(
+    def _reset_parameters_(self):  # noqa: D102
+        super()._reset_parameters_()
+        self.normal_vector_embeddings.reset_parameters()
+        # TODO: Add initialization
```
can we clarify what initialization is necessary here?

mberr replied:
This TODO was already existent before, so I am not 100% sure why it is there in the first place.
I suppose it is about explicitly initialising the `normal_vector_embeddings`, in case the pytorch default would change in the future.

cthoyt commented on src/pykeen/models/unimodal/trans_h.py:
```diff
@@ -90,47 +90,39 @@ def __init__(
-            self.entity_embeddings.weight,
-            self.normal_vector_embeddings.weight,
-            self.relation_embeddings.weight,
+            self.entity_embeddings(indices=None),
```
On first look, this seems a bit misleading but then i re-read the signature of `regularize_if_necessary()` and it's only reading some tensors, and not writing to them. What's the FIXME though?

mberr replied:
I am not sure why there is a FIXME 😅

cthoyt commented on src/pykeen/models/unimodal/trans_r.py:
```diff
@@ -81,38 +81,33 @@ def __init__(
+            entity_initializer=xavier_uniform_,
+            entity_constrainer=clamp_norm,
+            entity_constrainer_kwargs=dict(maxnorm=1., p=2, dim=-1),
+            relation_initializer=xavier_uniform_normed_,
```
should we dispense with `xavier_uniform_normed_` and just use the `compose_` function to create it on the fly?

mberr replied:
Yes, I think this would be easier.

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -0,0 +1,173 @@
+        """Apply constraints which should not be included in gradients."""
+
+
+# FIXME: It has to be more than one type...
```
can you clarify?

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -0,0 +1,173 @@
+    def post_parameter_update(self):  # noqa: D102
+        # apply constraints in-place
+        if self.constrainer is not None:
+            self._embeddings.weight.data = self.constrainer(self._embeddings.weight.data)
```
why is the constrainer an out-of-place function while all of the other ones are in-place functions?

mberr replied:
It allows using the built-in methods like functional.normalize without having to create in-place operations, which effectively would do the same,  `tensor.data = constrain(tensor.data)`)

cthoyt replied:
That makes sense - now looking at the code this is pretty obvious. Was it the other way before?

cthoyt commented on tests/test_models.py:
```diff
@@ -770,108 +819,7 @@ def _check_constraints(self):
         for emb in (self.model.entity_embeddings, self.model.relation_embeddings):
-            assert all_in_bounds(emb.weight.norm(p=2, dim=-1), high=1., a_tol=_EPSILON)
-
-    def test_score_hrt(self):
```
why get rid of the custom tests?

mberr replied:
re-introduced. renamed to avoid masking the generic `test_score_hrt` test

cthoyt committed:
Cosmetic improvements and remove code graveyard

cthoyt commented on src/pykeen/models/unimodal/rotate.py:
```diff
@@ -68,43 +98,12 @@ def __init__(
             random_seed=random_seed,
             regularizer=regularizer,
+            entity_initializer=xavier_uniform_,
+            relation_initializer=init_phases,
```
should `init_phases` be an in-place function?

cthoyt replied:
Possible implementation:

```python
def init_phases_(x: torch.Tensor):
    """Initialize with random phases in place using :func:`init_phases`."""
    x.data = init_phases(x)
    return x
```

mberr replied:
Or, we can make sure that `reset_parameters` of `Embedding` does this wrapping for all init methods.

mberr replied:
```python
def reset_parameters():
    x.data = self.initializer(x.data)
```

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -0,0 +1,173 @@
+        super().__init__()
+
+        if initializer is None:
+            initializer = nn.init.normal_
```
does it make sense to have a default for this?

mberr replied:
This is the default of torch.nn.Embedding, cf. https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html#torch.nn.Embedding

cthoyt replied:
Okay!

cthoyt committed:
Update docs

[skip ci]

mberr committed:
Extend documentation

mberr committed:
Remove in-place variants

mberr committed:
remove xavier_uniform_normed_

mberr committed:
Remove todos

mberr committed:
fix line-too-long

mberr committed:
Fix docstring

mberr committed:
Remove unused imports

mberr committed:
Revert removal of manual tests for score_hrt

mberr committed:
Avoid overriding test_score_hrt with manual tests

mberr committed:
fix manual test_hrt for TransR

mberr committed:
fix manual test_hrt for TransD

cthoyt committed:
Fix issue in TransR test
